### DISCUSSION TITLE
Problem importing Scanorama: dependency sklearn.externals.six

### DISCUSSION THREAD

**Discussion by auesro** [brianhie/scanorama#103] [Q&A] [answered]
I have used Scanorama previously (more than 1 year ago) but today I am getting an error on import:
```
import scanorama
Traceback (most recent call last):

  File "/tmp/ipykernel_22000/104445913.py", line 1, in <module>
    import scanorama

  File "/home/auesro/anaconda3/envs/Scanpy/lib/python3.7/site-packages/scanorama/__init__.py", line 1, in <module>
    from .scanorama import *

  File "/home/auesro/anaconda3/envs/Scanpy/lib/python3.7/site-packages/scanorama/scanorama.py", line 16, in <module>
    from .t_sne_approx import TSNEApprox

  File "/home/auesro/anaconda3/envs/Scanpy/lib/python3.7/site-packages/scanorama/t_sne_approx.py", line 31, in <module>
    from sklearn.externals.six import string_types

ModuleNotFoundError: No module named 'sklearn.externals.six'
```
From what I can see, `sklearn.externals.six` was dropped on scikit-learn version 0.21 as shown [here](https://scikit-learn.org/stable/whats_new/v0.21.html?highlight=externals%20six#sklearn-externals).
I could downgrade `scikit-learn` to 0.21 I guess but thats borderline with Python 3.7 support...any other solutions?

**Comment by brianhie** [comment, marked as answer]
Hi @auesro, thanks for reaching out! This should have been fixed in the latest version of Scanorama. Could you trying doing
```
python -m pip install --upgrade scanorama
```
and see if the error persists?

**Comment by auesro** [comment]
Hi @brianhie,
Problem solved. Indeed updating scanorama takes care of the issue.
Only to notice, I got a warning from pip:
```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
spyder 5.1.5 requires intervaltree>=3.0.2, but you have intervaltree 2.1.0 which is incompatible.
```
This obviously only affects Spyder users, so I just updated intervaltree to 3.0.2 and will check if Scanorama can run. If possible, check this out.

A

**Comment by auesro** [comment]
Scanorama seems to run just fine using intervaltree-3.0.2. But I will note that, so far, I only used the `correct_scanpy` wrapper (just in case that's important info).

And regarding the use of `correct` vs `integrate` I have a question (let me know if I should be opening a new issue or move it somewhere else): is it recommended that I use both functions using only the variable genes? I know this is recommended for integration but cannot find info regarding correction. Moreover, if I do correction on the variable genes only, I will lose the rest of the genes for downstream differential expression analysis, etc...any recommendation here?

Thanks!

A

**Comment by brianhie** [comment]
Sorry for the late reply! Yes, you are right that using variable genes is recommended for integration, though Scanorama integration is recommended only for constructing the KNN graph for visualization and clustering. I think the consensus in the field is to be very conservative in estimating differential expression, e.g., using the original count data or a simple method (e.g., Combat batch correction) for this purpose.
